platform/core/PolyPodCoreSwift/Sources/PolyPodCoreSwift/Codable/MessagePackEncoder.swift

```diff
@@ -0,0 +1,305 @@
+import PolyPodCore
+import Foundation
+import MessagePack
+
+public class MessagePackEncoder  {
+    public func encode<T>(_ value: T) throws -> MessagePackValue where T : Encodable {
+        if T.self == Data.self {
+            return .binary(value as! Data)
+        } else if T.self == Date.self {
+            return .string(ISO8601DateFormatter().string(from: value as! Date))
+        }
+        let encoder = _MessagePackEncoder()
+        try value.encode(to: encoder)
+        return encoder.value
+    }
+}
+
+protocol MessagePackEncodingContainer {
+    var value: MessagePackValue { get }
+}
+
+class SingleValueContainer: SingleValueEncodingContainer {
+    var codingPath: [CodingKey]
+    var userInfo: [CodingUserInfoKey: Any]
+    
+    init(codingPath: [CodingKey],
+         userInfo: [CodingUserInfoKey : Any]) {
+        self.codingPath = codingPath
+        self.userInfo = userInfo
+    }
+    
+    private var storage: MessagePackValue = MessagePackValue()
+    fileprivate var canEncodeNewValue = true
+    
+    fileprivate func checkCanEncode(value: Any?) throws {
+        guard self.canEncodeNewValue else {
+            let context = EncodingError.Context(
+                codingPath: self.codingPath,
+                debugDescription: "Cannot encode multiple values")
+            throw EncodingError.invalidValue(value as Any, context)
+        }
+    }
+    
+    func encodeNil() throws {
+        try checkCanEncode(value: nil)
+        defer { self.canEncodeNewValue = false }
+        storage = .nil
+    }
+    
+    func encode<T>(_ value: T) throws where T : Encodable {
+        try checkCanEncode(value: value)
+        defer { self.canEncodeNewValue = false }
+        
+        if T.self == Bool.self {
```

Comment:
Wouldn't `switch`/`case` work here?

Reply:
Unfortunately not because of some type system quirk. I get this warning message:

`Cast from 'T.Type' to unrelated type 'Bool' always fails`

Reply:
But you can do this:
```swift
switch value {
        case let bool as Bool:
            storage = .bool(bool)
        case let string as String:
            storage = .string(string)
        ....
}
```

Reply:
Ha, did not know about this. Thanks!

Reply:
Done